### PR TITLE
Create a hookable location checking function for actions

### PR DESCRIPTION
**Pull request by duncanspumpkin**
work in progress but this is so that you don't need to do really complex things to work out in a script if an action is working on a tile.

### CODE REVIEW

File: src/openrct2/scripting/HookEngine.h [hunk 1]
```diff
@@ -37,6 +37,7 @@ namespace OpenRCT2::Scripting
         NETWORK_JOIN,
         NETWORK_LEAVE,
         RIDE_RATINGS_CALCULATE,
+        GAME_ACTION_LOCATION,
```
Comment (IntelOrca) [inline src/openrct2/scripting/HookEngine.h:40]: I suggest removing `GAME_` so it is consistent with the other types which are just `ACTION_*`.

**Comment by IntelOrca** [comment]
`map_is_location_valid` checks if it is actually a valid tile on the map. We probably don't even want to call the hook unless this is true right? I'd then perhaps set `result` to the normal checks, like whether the land is owned. And then the script can check if result is true, and override that to false if it isn't allowed.

Is this something to be called for both query and execute? How do we handle auto clearance removal, e.g. like if a tree is in the way but gets automatically removed.

**Comment by duncanspumpkin** [comment]
I was imagining this being called in the query and the execute. Yes agree that the hook isn't worth calling if the check fails.

**Comment by duncanspumpkin** [comment]
I've used the check on if location valid as that will be called on all tiles that are being modified. If you don't have permission due to the hook it would not allow that tile to be modified. Auto scenery removal wouldn't be an issue as the function would return false preventing that from happening.

File: distribution/openrct2.d.ts [hunk 2]
```diff
@@ -350,6 +351,15 @@ declare global {
         nausea: number;
     }
 
+    interface ActionLocationArgs {
+        readonly x: number;
+        readonly y: number;
+        readonly playerId: number;
```
Comment (IntelOrca) [inline distribution/openrct2.d.ts:357]: The other network args just uses `player`.

Comment (duncanspumpkin) [reply distribution/openrct2.d.ts:357]: player for player Id?

**Pull request merged**